src/pudl_archiver/archivers/ferc/ferccid.py

```diff
@@ -0,0 +1,169 @@
+            download_path = (
+                self.download_directory / f"ferccid-{year}-{month}-{day}.csv"
+            )
+            yield self.download_cid_csv_by_paging_api(
+                SOURCE_URL, download_path=download_path
+            )
+
+    async def get_last_updated_date(self, page_url: str) -> str:
+        """Get the Data Last Updated date from the FERC data viewer page."""
+        soup = await self.get_soup(page_url)
+        # Find the element that contains the label text
+        label_div = soup.find(
+            lambda t: (
+                isinstance(t, Tag)
+                and t.name == "div"
+                and t.get_text(strip=True) == "Data Last Updated"
+            )
+        )
+        if not label_div:
+            raise RuntimeError("Couldn't find 'Data Last Updated' label div")
+
+        row = label_div.find_parent("div", class_="row")
+        if not row:
+            raise RuntimeError("Couldn't find parent row for 'Data Last Updated'")
+
+        # extract all text from row and remove the label itself
+        row_text = (
+            row.get_text(" ", strip=True).replace("Data Last Updated", "").strip()
+        )
+
+        match = DATE_LAST_UPDATED_PATTERN.search(row_text)
+        if not match:
+            raise RuntimeError(
+                f"Couldn't find date-like value near 'Data Last Updated'. Row text was: {row_text!r}"
+            )
+
+        return match
+
+    async def download_cid_csv_by_paging_api(
+        self,
+        page_url: str,
+        download_path: Path,
+        timeout_ms: int = 60_000,
+        page_size: int = 100,
+    ) -> ResourceInfo:
+        """Make request for CID data from inside the page and write to CSV."""
+        columns = [
+            "Organization_Name",
+            "CID",
+            "Program",
+            "Company_Website",
+            "Address",
+            "Address2",
+            "City",
+            "State",
+            "Zip",
+        ]
+
+        async with async_playwright() as pw:
+            browser = await pw.webkit.launch(headless=True)
+            page = await browser.new_page()
+            try:
+                await page.goto(
+                    page_url, wait_until="domcontentloaded", timeout=timeout_ms
+                )
+                await page.wait_for_load_state("networkidle", timeout=timeout_ms)
+
+                all_rows: list[dict[str, Any]] = []
+                start = 0
+                total_count: int | None = None
+
+                while True:
+                    payload = {
+                        "startRow": start,
+                        "endRow": start + page_size,
+                        "sortModel": [],
+                        "filterModel": {},
+                        "columns": columns,
+                        "castData": [],
+                    }
+                    # make the request from inside the page
+                    result = await page.evaluate(
```

Comment:
Could you add some clarifying comments about what's going on here (in the docstring and/or inline)? From what I can tell we're continuously querying the actual api that supplies the data for the interactive table on the CID page until we've fetched all rows, but it would be nice to have a bit more clarification on exactly what we're doing/why.

Also, is there a reason we have to use `playwright` to query the API from within the page using Javascript? could we just query the API directly with python?

Reply:
It didn't work for me to query the API directly with python. I don't fully understand why, but it seems like you need to simulate being in a browser before making the request. I'll leave some comments.